src/error.rs

```diff
@@ -168,7 +168,34 @@ pub enum ProtocolError {
 
     /// Raised when serializing or de-serializing a message body fails.
     #[fail(display = "ProtocolError: serialization error ({})", _0)]
-    BodySerializationError(#[fail(cause)] serde_json::Error),
+    BodySerializationError(#[fail(cause)] FormatError),
+}
+
+#[derive(Debug, Fail)]
+pub enum FormatError {
+    #[fail(display = "{}", _0)]
+    Json(serde_json::Error),
+
+    #[cfg(any(test, feature = "extra_formats"))]
+    #[fail(display = "{}", _0)]
+    Yaml(serde_yaml::Error),
+
+    #[cfg(any(test, feature = "extra_formats"))]
+    #[fail(display = "{}", _0)]
+    Pickle(serde_pickle::error::Error),
+
+    #[cfg(any(test, feature = "extra_formats"))]
+    #[fail(display = "{}", _0)]
+    MsgPackDecode(rmp_serde::decode::Error),
+    #[cfg(any(test, feature = "extra_formats"))]
+    #[fail(display = "{}", _0)]
+    MsgPackEncode(rmp_serde::encode::Error),
```

Comment:
More space nitpicking

```suggestion
    MsgPackEncode(rmp_serde::encode::Error),
    
```